### PR TITLE
mattermost: Refactoring and upgrades

### PR DESCRIPTION
**Pull request by Kranzes**
###### Description of changes
I migrated the `mattermost` package from buildGoPackage to buildGoModule which is now the standard. 
The package expression has not been worked on (excluding updates) in about 5 years. 
- [mattermost: 6.3.6 -> 7.0.0](https://github.com/NixOS/nixpkgs/pull/178389/commits/21738df66b20c274118395cd63460097b0a3e8fd)
- [mattermost: 7.0.0 -> 7.0.1](https://github.com/NixOS/nixpkgs/pull/178389/commits/c750d488035c9dc234523f3b5e62575cb784124c)
- [mmctl: 6.4.2 -> 7.0.0](https://github.com/NixOS/nixpkgs/pull/178389/commits/ad439443f11aeeff9a63171b604bd8dc6d78321d)
- [mmctl: 7.0.0 -> 7.0.1](https://github.com/NixOS/nixpkgs/pull/178389/commits/09724b15170081415c1559dc4508584b4de61b42)

Closes #167281

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Kranzes** [comment]
Result of `nixpkgs-review pr 178389` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>mattermost</li>
    <li>mmctl</li>
  </ul>
</details>

**Comment by Kranzes** [comment]
Result of `nixpkgs-review pr 178389` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>mattermost</li>
    <li>mmctl</li>
  </ul>
</details>

File: pkgs/servers/mattermost/default.nix [hunk 1]
```diff
@@ -1,101 +1,55 @@
...
-        for arg in "$@"; do
-          if [[ "$arg" == -ldflags=* ]] && [ $found -eq 0 ]; then
-            arg="-ldflags=''${ldflags} ''${arg#-ldflags=}"
-            found=1
-          fi
-          args+=("$arg")
-        done
-        "$origGo" "''${args[@]}"
-      }
+  subPackages = [ "cmd/${pname}" ];
```
Comment (SuperSandro2000) [inline pkgs/servers/mattermost/default.nix:26]:
```suggestion
  subPackages = [ "cmd/mattermost" ];
```

**Review by SuperSandro2000 — COMMENTED** [review]
otherwise LGTM

**Comment by Kranzes** [comment]
Result of `nixpkgs-review pr 178389` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>mattermost</li>
    <li>mmctl</li>
  </ul>
</details>

**Review by MatthewCroughan — APPROVED** [review]

File: pkgs/servers/mattermost/default.nix [hunk 2]
```diff
@@ -1,101 +1,51 @@
...
 
-      # For some reason a bunch of these files are +x...
-      find $out -type f -exec chmod -x {} \;
-    '';
+  meta = with lib; {
+    description = "Mattermost is an open source platform for secure collaboration across the entire software development lifecycle";
+    homepage = "https://www.mattermost.org";
+    license = with licenses; [ agpl3 asl20 ];
+    maintainers = with maintainers; [ fpletz ryantm numinit kranzes ];
+    platforms = [ "x86_64-linux" ];
```
Comment (fpletz) [inline pkgs/servers/mattermost/default.nix:49]: Why is this restriction necessary? The go code should work on every platform we support and the webapp should be architecture-agnostic, right?

Comment (Kranzes) [reply pkgs/servers/mattermost/default.nix:49]: You're right.

**Review by fpletz — APPROVED** [review]

**Comment by fpletz** [comment]
Thanks for the improvements!

**Comment by numinit** [comment]
Appreciate it!

**Pull request merged**